res/controllers/Hercules-DJControl-Inpulse-300-script.js

```diff
@@ -44,76 +60,239 @@ DJCi300.scratchShiftMultiplier = 4;
-    engine.connectControl("[Channel1]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-	engine.getValue("[Channel1]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-    engine.connectControl("[Channel2]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-	engine.getValue("[Channel2]", "vu_meter", "DJCi300.vuMeterUpdateDeck");
-    engine.connectControl("[Main]", "vu_meter_left", "DJCi300.vuMeterUpdateMaster");
-    engine.connectControl("[Main]", "vu_meter_right", "DJCi300.vuMeterUpdateMaster");
-	engine.getValue("[Main]", "vu_meter_left", "DJCi300.vuMeterUpdateMaster");
-    engine.getValue("[Main]", "vu_meter_right", "DJCi300.vuMeterUpdateMaster");
+    // Turn On Toneplay LED
+    midi.sendShortMsg(0x96, 0x40, 0x7F);
+    midi.sendShortMsg(0x96, 0x48, 0x7F);
+    midi.sendShortMsg(0x97, 0x40, 0x7F);
+    midi.sendShortMsg(0x97, 0x48, 0x7F);
+
+    //Softtakeover for Pitch fader
+    engine.softTakeover("[Channel1]", "rate", true);
+    engine.softTakeover("[Channel2]", "rate", true);
+    engine.softTakeoverIgnoreNextValue("[Channel1]", "rate");
+    engine.softTakeoverIgnoreNextValue("[Channel2]", "rate");
+
+    // Connect the VUMeters
+    engine.makeConnection("[Channel1]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Channel2]", "VuMeter", DJCi300.vuMeterUpdateDeck);
+    engine.makeConnection("[Master]", "VuMeterL", DJCi300.vuMeterUpdateMaster);
+    engine.makeConnection("[Master]", "VuMeterR", DJCi300.vuMeterUpdateMaster);
+
+    // Connect beatmatch LED functions
+    engine.makeConnection("[Channel1]", "bpm", DJCi300.updateBeatmatchTempoLED);
+    engine.makeConnection("[Channel2]", "bpm", DJCi300.updateBeatmatchTempoLED);
+    // We only connect channel 1 here because beatmatch is only enabled when both decks are connected anyway
+    // This saves resources
+    engine.makeConnection("[Channel1]", "beat_distance", DJCi300.updateBeatmatchAlignLED);
+    // We also want to update all beatmatch LEDs when a song is played
+    engine.makeConnection("[Channel1]", "play", DJCi300.updateBeatmatchAlignLED);
+    engine.makeConnection("[Channel2]", "play", DJCi300.updateBeatmatchAlignLED);
+    engine.makeConnection("[Channel1]", "play", DJCi300.updateBeatmatchTempoLED);
+    engine.makeConnection("[Channel2]", "play", DJCi300.updateBeatmatchTempoLED);
+
+    // Connect jogwheel functions
+    engine.makeConnection("[Channel1]", "scratch2", DJCi300.updateScratchAction);
+    engine.makeConnection("[Channel2]", "scratch2", DJCi300.updateScratchAction);
+
+    // Connect the toneplay LED updates
+    engine.makeConnection("[Channel1]", "pitch", DJCi300.updateToneplayLED);
+    engine.makeConnection("[Channel2]", "pitch", DJCi300.updateToneplayLED);
+
+    // Define slicer connections by connecting them
+    DJCi300.connectSlicerFunctions(1);
+    DJCi300.connectSlicerFunctions(2);
+    // Then disconnect them since we're not in slicer mode (yet)
+    DJCi300.disconnectSlicerFunctions(1);
+    DJCi300.disconnectSlicerFunctions(2);
```

Comment:
I think a better model here would be to have the functions always connected, and then use a bool to decide if they are active.  So the handler functions would say "if slicing is not active, return".  Then you don't need to be making and disconnecting connections all the time.

Reply:
That results in lots of function call overhead from C++ into the JS. I think we need to carefully consider how many of these are active vs disabled. (dis-)Connecting is not that expensive IIRC. its just a hashmap lookup with a `QObject::connect` call.